### ISSUE TITLE
Silent corruption of node.run_context

### ISSUE THREAD

**Issue by coderanger** [chef/chef#3485]
The new guard_interpreter system silently corrupts `node.run_context` and no one has noticed for a year.

The problematic line is in [`ResourceGuardInterpreter#get_interpreter_resource`](https://github.com/chef/chef/blob/master/lib/chef/guard_interpreter/resource_guard_interpreter.rb#L96):

```ruby
anonymous_run_context = Chef::RunContext.new(parent_resource.node, {}, empty_events)
```

This would be fine except that in [`RunContext#initialize`](https://github.com/chef/chef/blob/master/lib/chef/run_context.rb#L88) we do this:

```ruby
@node.run_context = self
```
This means that after any resource runs a guard where `guard_interpreter` is not `:default`, `node.run_context` is set to that fake run context. This then breaks some code in Poise that tries to use `node.run_context.cookbook_collection` and finds it has been set to `{}`.

**Comment by coderanger** [comment]
Ping @adamedx @danielsdeleo @lamont-granquist.

**Comment by coderanger** [comment]
A minimal repro case:

```ruby
log 'before' do
  message lazy { "before = #{run_context.cookbook_collection.object_id} #{node.run_context.cookbook_collection.object_id}" }
  level :error
end

execute 'true' do
  not_if 'true'
end

log 'after' do
  message lazy { "after = #{run_context.cookbook_collection.object_id} #{node.run_context.cookbook_collection.object_id}" }
  level :error
end
```

**Comment by adamedx** [comment]
Whoa, we set the private state of another object? Yuk. Quick fix would then be a new static method on `RunContext` that could safely generate an instance of the class without messing with another object's state.

**Comment by coderanger** [comment]
That's an option, or removing that from `RunContext#initialize`. A more immediate fix would probably be to do like Poise and use_inline_resources do and `parent_resource.run_context.dup` rather than making a new one from scratch. You'll have to create a new resource collection just like use_inline_resources does though. https://github.com/chef/chef/blob/master/lib/chef/provider/lwrp_base.rb#L56-L58

Long term making the initialize not set semi-global state seems like a better fix, I bet there are other instances of this lurking in weird places.

**Comment by adamedx** [comment]
Yeah, I was thinking of the dup approach just now actually. First just getting my spec to fail -- we have unit tests that are not making it easy for me to hit this, and the functional tests are all bound up in specific resources.

**Comment by coderanger** [comment]
Test that is failing for me:

```ruby
    it "does not change global state in node.run_context" do
      starting_run_context_id = run_context.object_id
      parent_resource.guard_interpreter(:script)
      guard_interpreter.send(:get_interpreter_resource, parent_resource)
      expect(node.run_context.object_id).to eq(starting_run_context_id)
    end
```

**Comment by adamedx** [comment]
Fixed a mistake in the unit spec, now I have a repro.

**Comment by adamedx** [comment]
Ah, yeah just basically wrote your same test @coderanger :smile: .

**Comment by adamedx** [comment]
Here is work in progress: https://github.com/chef/chef/compare/adamedx/3485-run_context_corruption .

**Comment by coderanger** [comment]
For the hell of it, this makes the test pass:

```diff
diff --git a/lib/chef/guard_interpreter/resource_guard_interpreter.rb b/lib/chef/guard_interpreter/resource_guard_interpreter.rb
index 1e2a534..93dff50 100644
--- a/lib/chef/guard_interpreter/resource_guard_interpreter.rb
+++ b/lib/chef/guard_interpreter/resource_guard_interpreter.rb
@@ -93,7 +93,9 @@ class Chef
         end

         empty_events = Chef::EventDispatch::Dispatcher.new
-        anonymous_run_context = Chef::RunContext.new(parent_resource.node, {}, empty_events)
+        anonymous_run_context = parent_resource.run_context.dup
+        anonymous_run_context.resource_collection = Chef::ResourceCollection.new
+        anonymous_run_context.instance_variable_set(:@events, empty_events)
         interpreter_resource = resource_class.new('Guard resource', anonymous_run_context)
         interpreter_resource.is_guard_interpreter = true
```

Probably not a great solution given the `instance_variable_set`, but its related to a solution I bet.

**Comment by adamedx** [comment]
@coderanger, thank you for the excellent find. We've merged #3497, so Poise should be good to go on master now.
